### PR TITLE
Loop for newest version in appstore response

### PR DESCRIPTION
**Pull request by LukasReschke**
The current implementation when fetching apps from the appstore is to assume that the first element is the newest version, this is now always applicable (e.g. daily builds etc.) and leads to the fact that for some apps (e.g. nextant) the newest version is not delivered. This can be easily tested by comparing the version of the downloaded Nextant version, without this patch it only delivers 0.6.6 instead of 0.10.0. (Category: Files)

This change will loop over all releases delivered by the appstore and chooses the newest compatible one. While not the cleanest solution, it does its job.

Most of the code are actually unit tests. Whereas I have copied the whole original response from the appstore and also have performed the transformation. So that's why the diff looks so huge.

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

### CODE REVIEW

**Comment by nickvergessen** [comment]
Works, afterwards i am suggested an update to 0.10.0 on the category page, normal enabled page however does not.

**Comment by LukasReschke** [comment]
> normal enabled page however does not.

Good point. Adjusted the logic to also cover that. This controller file is really a little bit … special …

**Comment by nickvergessen** [comment]
Works now 👍

**Comment by blizzz** [comment]
Code looks OK, works 👍

**Comment by BernhardPosselt** [comment]
Hm, this does not filter pre-releases, is this what you want? Users will be prompted to upgrade to 2.0.0-beta.1 for instance

**Comment by LukasReschke** [comment]
> Hm, this does not filter pre-releases, is this what you want? Users will be prompted to upgrade to 2.0.0-beta.1 for instance

https://github.com/nextcloud/server/pull/2307/files#diff-83840d74b03720feef2006e83453333eR83 should remove all nightlies. Am I missing something?

File: lib/private/App/AppStore/Fetcher/AppFetcher.php [hunk 1]
```diff
@@ -53,4 +58,66 @@ public function __construct(IAppData $appData,
...
+		$ncMajorVersion = explode('.', $ncVersion)[0];
+		foreach($response['data'] as $dataKey => $app) {
+			$releases = [];
+
+			// Filter all compatible releases
+			foreach($app['releases'] as $release) {
+				// Exclude all nightly releases
+				if($release['isNightly'] === false) {
+					// Exclude all versions not compatible with the current version
+					$versionParser = new VersionParser();
```
Comment (BernhardPosselt) [inline lib/private/App/AppStore/Fetcher/AppFetcher.php]: This should not be needed. All apps fetched from the API are working on the version provided in your URL

**Comment by BernhardPosselt** [comment]
@LukasReschke right, however there are 3 "channels":
* stable (e.g. 1.0.0)
* pre-release (e.g. 1.0.0-beta)
* nightly (e.g. 1.0.0-beta + nightly flag)

**Comment by LukasReschke** [comment]
So for what does one have to filter? The docs doesn't tell anything about pre-releases etc…

(also it seems rather unnecessary for me at the moment since Nextcloud only supports the stable channel at the moment, but fair enough…)

**Comment by BernhardPosselt** [comment]
Pre-releases in semver are denoted by having a - after the version string. Since we don't allow build metadata in our version you can simply check with 

```php

if (strpos($release['version'], '-') === false) {
    // not a pre-release
}

```

**Comment by BernhardPosselt** [comment]
Versioning is documented at http://nextcloudappstore.readthedocs.io/en/latest/restapi.html#get-all-apps-and-releases 👍

version
    A semantic version without build metadata (e.g. 1.3.0, 1.2.1-alpha.1)

If you think this was not enough docs to handle this case, can you create a pull request to the docs that specifically adds a short description about pre-releases :D?

**Comment by nickvergessen** [comment]
@BernhardPosselt Ever heard of -PL releases? just saying

**Comment by LukasReschke** [comment]
https://github.com/nextcloud/server/pull/2327 should make you all happier… I hope…

**Comment by BernhardPosselt** [comment]
@nickvergessen nope, havent heard of that one

**Pull request merged**